src/storage/src/wal.rs

```diff
@@ -0,0 +1,385 @@
+        self.wals.insert(id, wal);
+        Ok(())
+    }
+
+    pub fn destroy(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if !self.wals.contains_key(&id) {
+            bail!(
+                "asked to delete wal for relation: {:?} that doesn't exist",
+                id
+            )
+        }
+
+        self.wals.remove(&id);
+        Ok(())
+    }
+
+    pub fn write(&mut self, id: GlobalId, updates: &[Update]) -> Result<(), anyhow::Error> {
+        if !self.wals.contains_key(&id) {
+            bail!("asked to write to unknown relation: {}", id)
+        }
+
+        let wal = self.wals.get_mut(&id).expect("wal known to exist");
+        wal.write(updates)?;
+        Ok(())
+    }
+
+    pub fn resume(&mut self, id: GlobalId) -> Result<(), anyhow::Error> {
+        if self.wals.contains_key(&id) {
+            bail!(
+                "asked to resume wal for relation: {:?} that already exists.",
+                id
+            )
+        }
+        let wal_path = self.get_path(id);
+        let wal = WriteAheadLog::resume(id, wal_path)?;
+        self.wals.insert(id, wal);
+
+        Ok(())
+    }
+
+    pub fn write_progress(&mut self, timestamp: Timestamp) -> Result<(), anyhow::Error> {
+        // For all logs, indicate that <timestamp> is now the min timestamp going
+        // forward
+        for (_, wal) in self.wals.iter_mut() {
+            wal.write_progress(timestamp)?;
+        }
+
+        Ok(())
+    }
+}
+
+// Write a 20 byte header + length-prefixed Row to a buffer
+// TODO: change the on-disk representation to be more compact.
+pub fn encode_update(
+    row: &Row,
+    timestamp: Timestamp,
+    diff: isize,
+    buf: &mut Vec<u8>,
+) -> Result<(), anyhow::Error> {
+    assert!(diff != 0);
+    assert!(timestamp > 0);
+    let data = row.data();
+
+    if data.len() >= u32::MAX as usize {
+        bail!("failed to encode row: row too large");
+    }
+
+    // Write out the header
+    // Not a progress message so - 0
+    buf.write_u32::<NetworkEndian>(0).unwrap();
+    buf.write_u64::<NetworkEndian>(timestamp).unwrap();
+    buf.write_i64::<NetworkEndian>(diff as i64).unwrap();
+
+    // Now write out the data
+    buf.write_u32::<NetworkEndian>(data.len() as u32)
+        .expect("writes to vec cannot fail");
+    buf.extend_from_slice(data);
+    Ok(())
+}
+
+pub fn encode_progress(timestamp: Timestamp, buf: &mut Vec<u8>) -> Result<(), anyhow::Error> {
+    // This is a progress update so - 1
+    buf.write_u32::<NetworkEndian>(1).unwrap();
+    buf.write_u64::<NetworkEndian>(timestamp).unwrap();
+
+    Ok(())
+}
+
+// Open a new log segment file to write into. Importantly, we need to open this
+// file with O_APPEND and we should fail if the log segment file already exists.
+fn create_log_segment(path: &Path) -> Result<File, anyhow::Error> {
+    fs::OpenOptions::new()
+        .append(true)
+        .create_new(true)
+        .open(path)
```

Comment:
It seems that a file is being opened and kept open for each relation, which can lead to resource exhaustion. While MZ does seem to call ```prlimit64``` to get itself a 1M open-file-limit, I think this is going to become a problem in the future.

Core banking applications are known to have tens of thousands of base tables. Even if it will not make sense for them to be created in MZ, I can totally envison a situation where a customer will attempt to create an unreasonable amount of tables in MZ, even if they do not intend to populate/query them.

Reply:
I can document this behavior in the comments and add a todo to resolve. totally makes sense that this could be a future pain point

Reply:
I am not sure how TODOs are tracked, maybe opening a ticket would increase the probability that this is not forgotten.

Reply:
I'm documenting this in a TODO -- for now I think thats better because I think this code will have quite a bit of churn in the next 2 - 3 weeks so I want to let that settle and then I'll file all the issues that remain. hope that sounds ok feel free to unresolve this thread if not!